File: junction/samples/serverless/http/watchdog.cc
#include "watchdog.h"

#include <sys/stat.h>
#include <unistd.h>

#include <iostream>
#include <string>

#include "lib/httplib.h"

constexpr const char *SOCK_DIR = "/tmp/serverless/";
constexpr const char *SOCK_EXT = ".sock";

WatchDog::WatchDog(const std::string &name, const std::string &req_path,
                   httplib::Server::Handler h)
    : handler_(std::move(h)) {
  sock_path_ = SOCK_DIR + name + SOCK_EXT;
  req_path_ = req_path;
}

bool WatchDog::InitServer() {
  struct stat st = {0};
  if (stat(SOCK_DIR, &st) == -1) {
    if (mkdir(SOCK_DIR, 0777) != 0 && errno != EEXIST) {
      std::cerr << "[Watchdog] Failed to create directory: " << SOCK_DIR
                << std::endl;
      return false;
    }
  }
  unlink(sock_path_.c_str());

  httplib::Server svr;
  svr.Get(req_path_, handler_);

  if (!svr.set_address_family(AF_UNIX).listen(sock_path_, 80)) {
    std::cerr << "[Watchdog] Failed to listen on " << sock_path_ << std::endl;
    return false;
  }

  return true;
}

void WatchDog::Run() {
  if (!InitServer()) { exit(1); }
}

File: junction/samples/serverless/http/watchdog.py
import os
import socket
import sys
from http.server import HTTPServer, BaseHTTPRequestHandler

SOCK_DIR = "/tmp/serverless/"
SOCK_EXT = ".sock"

class UnixHTTPServer(HTTPServer):
    address_family = socket.AF_UNIX

    def server_bind(self):
        # Create the directory if it doesn't exist
        if not os.path.exists(SOCK_DIR):
            try:
                os.makedirs(SOCK_DIR, mode=0o777)
            except OSError as e:
                print(f"[Watchdog] Failed to create directory: {e}", file=sys.stderr)
                sys.exit(1)

        # Unlink the socket file if it exists
        if os.path.exists(self.server_address):
            os.unlink(self.server_address)

        # Bind the socket
        HTTPServer.server_bind(self)
        
        # Set permissions so others can write/connect to it
        os.chmod(self.server_address, 0o777)

class WatchDog:
    def __init__(self, name, handler_class):
        self.sock_path = os.path.join(SOCK_DIR, name + SOCK_EXT)
        self.handler_class = handler_class

    def run(self):
        try:
            # Create the server bound to the Unix Socket path
            server = UnixHTTPServer(self.sock_path, self.handler_class)
            print(f"[Watchdog] Listening on {self.sock_path}...")
            server.serve_forever()
        except Exception as e:
            print(f"[Watchdog] Critical Error: {e}", file=sys.stderr)
            sys.exit(1)
